bindings/python/src/lib.rs

```diff
@@ -441,9 +458,166 @@ struct Open {
     storage: Arc<Storage>,
 }
 
+const ALIGN: usize = 4096; // TODO: use ioctl to determine correct alignment size
+const CHUNK_SIZE: usize = 1 << 20; // 1 mib
+
 impl Open {
-    fn new(filename: PathBuf, framework: Framework, device: Option<Device>) -> PyResult<Self> {
-        let file = File::open(&filename).map_err(|_| {
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    fn read_odirect(filename: impl AsRef<Path>) -> PyResult<DirectIo> {
+        let filename = filename.as_ref().to_str().ok_or_else(|| {
+            SafetensorError::new_err(format!(
+                "Path {} is not valid UTF-8",
+                filename.as_ref().display()
+            ))
+        })?;
+
+        unsafe {
+            let c_path = std::ffi::CString::new(filename)
+                .map_err(|_| SafetensorError::new_err("Failed converting file path to CString"))?;
+            let fd = libc::open(c_path.as_ptr(), libc::O_RDONLY | libc::O_DIRECT);
+            if fd < 0 {
+                return Err(SafetensorError::new_err(format!(
+                    "Could not open file {filename} with O_DIRECT",
+                )));
+            }
+            let file = File::from_raw_fd(fd);
+            let meta = file.metadata().map_err(|_| {
+                SafetensorError::new_err(
+                    format!("Could not retrieve metadata for file {filename}",),
+                )
+            })?;
+            let file_len = meta.len() as usize;
+            std::mem::forget(file);
+            if file_len == 0 {
+                libc::close(fd);
+                return Ok(DirectIo::from_vec(vec![]));
+            }
+
+            let aligned_prefix_len = file_len / ALIGN * ALIGN;
+            let tail_len = file_len - aligned_prefix_len;
+
+            let aligned_capacity = file_len.checked_next_multiple_of(ALIGN).ok_or_else(|| {
+                SafetensorError::new_err("overflow while computing aligned file len")
+            })?;
+            let layout = Layout::from_size_align(aligned_capacity, ALIGN)
+                .map_err(|_| SafetensorError::new_err("invalid layout"))?;
+            let ptr = alloc(layout);
+            if ptr.is_null() {
+                handle_alloc_error(layout);
+            }
+            let mut vec: Vec<u8> = Vec::from_raw_parts(ptr, 0, aligned_capacity);
```

Comment:
I might be mistaken, but I think `Vec` uses the alignment of the type, so you can do this more safely using something like:

```
let mut vec: Vec<[u8; ALIGN]> = Vec::with_capacity((file_len + ALIGN - 1) / ALIGN);
// Use vec as u8 rather than [u8; ALIGN].
```